client/asset/eth/eth.go

```diff
@@ -2371,6 +2462,119 @@ func (eth *ETHWallet) checkForNewBlocks(reportErr func(error)) {
 			w.checkFindRedemptions()
 		}
 	}()
+	go func() {
+		for _, w := range eth.walletList() {
+			w.checkConfirmRedemptions()
+		}
+	}()
+}
+
+func (w *assetWallet) swapIsRedeemed(secretHash [32]byte, contractVersion uint32) (bool, error) {
+	swap, err := w.swap(w.ctx, secretHash, contractVersion)
+	if err != nil {
+		return false, err
+	}
+	return swap.State == dexeth.SSRedeemed, nil
+}
+
+// confirmRedemption checks one of the queued redemption requests.
+func (w *assetWallet) confirmRedemption(txHash common.Hash, req *confirmRedemptionRequest, currentTip int64) {
+	const (
+		txConfsNeededToConfirm               = 10
+		blocksToWaitBeforeCoinNotFound       = 5
+		blocksToWaitBeforeCheckingIfReplaced = 10
+	)
+
+	var blocksSinceRequestStart int64
+	if currentTip > req.startHeight {
+		blocksSinceRequestStart = currentTip - req.startHeight
+	}
+
+	tx, txBlock, err := w.node.getTransaction(w.ctx, txHash)
+	if err == asset.CoinNotFoundError {
+		w.log.Errorf("confirmRedemption for tx %s: could not find coin", txHash)
+		// Wait a few blocks before returning that the coin cannot be found.
+		if blocksSinceRequestStart < blocksToWaitBeforeCoinNotFound {
+			return
+		}
+		redeemed, redeemedErr := w.swapIsRedeemed(req.secretHash, req.contractVersion)
+		if redeemedErr != nil {
+			req.res <- redeemedErr
+			return
+		}
+		if redeemed {
+			req.res <- nil
+			return
+		}
+	}
+	if err != nil {
+		req.res <- err
+		return
+	}
+
+	var confirmations int64
+	if txBlock > 0 {
+		if txBlock > currentTip {
+			confirmations = 1
+		} else {
+			confirmations = currentTip - txBlock + 1
+		}
+	}
+
+	if confirmations >= txConfsNeededToConfirm {
+		req.res <- nil
+		return
+	} else if confirmations > 0 {
+		return
+	}
+
+	// If the transaction is still unconfirmed, but was submitted less than 10
+	// blocks ago, give it some time before checking if it has been replaced,
+	// or redeemed by another tx.
+	if blocksSinceRequestStart%blocksToWaitBeforeCheckingIfReplaced != 0 || blocksSinceRequestStart == 0 {
+		return
+	}
+
+	redeemed, err := w.swapIsRedeemed(req.secretHash, req.contractVersion)
+	if err != nil {
+		req.res <- err
+		return
+	}
+	if redeemed {
+		req.res <- nil
+		return
+	}
```

Comment:
> The tx could be a lie.

What do you mean here?

Reply:
We don't check that the tx actually redeemed the contract. Do we?

I guess we would just need to check the tx data and make sure it's a redemption for the right secret hash to the right contract and the tx status to see if it succeeded altering the contract? I'm not sure if that is enough but maybe is.

Say the redeem for some reason had a tx here that was mined but failed to redeem. Then we made another tx that (not with this code) did redeem successfully before this method happens. We would be looking at failed tx confirms which is not when the contract was redeemed.

It's fine here because it's our redeem. I just don't want the same thing to be done for a conterparty's tx in the future so I think it's NOTEworthy.

Reply:
> You would need to add a `redeemBlockNumber` to the contract's swap struct.

We could check the state of the contract at 10 blocks before the tip as well. I was thinking about doing that, but I thought it wasn't necessary.

I'll add a note though.

Reply:
Thanks.